lib/runtime/printk.c

```diff
@@ -219,17 +178,53 @@ struct out_context {
 	int count;
 };
 
-static int char_out(int c, struct out_context *ctx)
+
+#ifdef CONFIG_USERSPACE
+K_SEM_DEFINE(printk_sem, 1, 1);
+
+struct printk_buffer {
+	int count;
+	char buf[CONFIG_PRINTK_BUFFER_SIZE];
+};
+
+static struct printk_buffer pbuf;
+
+static void buf_flush(void)
+{
+	_k_str_out(pbuf.buf, pbuf.count);
+	pbuf.count = 0;
+}
+
+static void char_out(int c, struct out_context *ctx)
 {
 	ctx->count++;
-	return _char_out(c);
+	pbuf.buf[pbuf.count++] = c;
+	if (pbuf.count == CONFIG_PRINTK_BUFFER_SIZE - 2) {
+		buf_flush();
+	}
 }
+#else
+static void char_out(int c, struct out_context *ctx)
+{
+	ctx->count++;
+	_k_char_out(c);
+}
+#endif
 
 int vprintk(const char *fmt, va_list ap)
 {
 	struct out_context ctx = { 0 };
 
+#ifdef CONFIG_USERSPACE
+	k_sem_take(&printk_sem, K_FOREVER);
```

Comment:
This semaphore does not work well in practice, as no permissions to threads have been granted for it.
So any user thread trying to printk() immediately faults from not having permission on this semaphore.
Need to think about this.

Reply:
Put the buffer inside this function when `CONFIG_USERSPACE`.  I don't see any issues here -- sure, stack usage will go slightly up temporarily, but if this feature is enabled, memory usage goes up already.

Since it's not a global variable, the space will only be used when printk() is being called -- which might be better than the global.

Not having a semaphore is a good thing as well, since each printk() call would be 3 system calls at minimum instead of just 1.

Reply:
Agreed. Working on this now.